### PR TITLE
Complete payment feature

### LINKED ISSUE #872: Add the option to buy Sharezone Plus on web platforms
Possible services that need to be elevated:

* Stripe
* [Paddle](https://www.paddle.com/)

### LINKED ISSUE #899: Add analytics for Sharezone Plus page
Analytics for:

* (un)subscribing Sharezone Plus
* opening an advantage
* opening a faq question
* opening privacy policy
* opening terms of terms

### LINKED ISSUE #932: Add feature flag to disable buying Sharezone Plus
## Description

It would be good to have the option to disable the purchase of Sharezone Plus for some possible future reasons. Reasons could be:

* We shut down Sharezone Plus
* Our servers are currently down
* We have a major payment problem

Not sure if we should implement this with Remote Config. Maybe a cloud function when clicking the subscribe button would be a better solution because it's immediately available.

## Open questions

* Should we block the payment if the available check couldn't be made (fail open vs. fail closed)?
  * Pro: If the cloud functions are down, the chances that Firebase is down is high. Therefore, buying Sharezone Plus would not be possible, which would be good.
  * Con: If for some unimportant the cloud function fails (but the cloud functions are healthy), a potential subscription could be missed.
  * I think we should fail closed and block the payment. However, we should use the `retry` method.

## Possible design of the cloud function:

Input:
* Platform: Android | iOS | Web | macOS (allows us to block the payment for a specific platform).
* App version: v1.3.7 (allows us to block payment for a specific version)

Answer:
* Status: Sharezone is available | Sharezone is _not_ available
* Error message (optional): e.g. "Buying Sharezone Plus is temporarily disabled. Please come back in a few hours", "We have stopped offering Sharezone Plus".

_Note: I think we should keep this system simple and not over-engineer it (it has a lot of potential to be over-engineered) because disabling the purchase of Sharezone Plus should not happen that often._

### LINKED ISSUE #1484: Add lifetime option for Sharezone Plus

### PR DESCRIPTION
**Pull request by nilsreichardt**
## Description

This PR (nearly) completes the payment feature of the Sharezone Plus subscription. The following things were added:

* Flag to disable new purchases (#932)
* Analytics for the subscription page (#899)
* Cancel subscription
* Add lifetime option for Sharezone Plus (#1484)
* Move buy section to `sharezone_plus_page_ui` package (will be used for "Eltern bezahlen lassen" feature)
* Dialogs for payment failure
* Update RevenueCat package to latest version

## Demo

https://github.com/SharezoneApp/sharezone-app/assets/24459435/bf6102f6-c190-44f7-bfb0-da05f5732fd5

(Used Stripe for the Demo, we will use RevenueCat in production for macOS)

## ToDo

* [ ] Allow to resubscribe when you cancelled subscription
* [ ] Create products in App Store
* [ ] Create products in Play Store
* [ ] Set `isCancelled` when Stripe subscription is cancelled via customer portal (something for the backend repo)
* [ ] Ensure Stripe invoice is sent for monthly & lifetime subscription  (something for the backend repo)

Closes #932
Closes #872
Closes #1484
Closes #899

### CODE REVIEW

File: app/lib/sharezone_plus/page/sharezone_plus_page_controller.dart [hunk 1]
```diff
@@ -105,13 +180,82 @@ class SharezonePlusPageController extends ChangeNotifier {
...
+
   @override
   void dispose() {
     _hasPlusSubscription?.cancel();
+    _sharezonePlusStatusSubscription?.cancel();
     super.dispose();
   }
 }
+
+class CanNotCancelOnThisPlatform implements Exception {
```
Comment (Jonas-Sander) [inline app/lib/sharezone_plus/page/sharezone_plus_page_controller.dart:249]:
```suggestion
class CanNotCancelOnThisPlatformException implements Exception {
```

File: app/lib/sharezone_plus/page/sharezone_plus_page_controller.dart [hunk 2]
```diff
@@ -105,13 +180,82 @@ class SharezonePlusPageController extends ChangeNotifier {
...
+  final SubscriptionSource? source;
+
+  const CanNotCancelOnThisPlatform(this.source);
+}
+
+class CouldNotGetManagementUrl implements Exception {
```
Comment (Jonas-Sander) [inline app/lib/sharezone_plus/page/sharezone_plus_page_controller.dart:255]:
```suggestion
class CouldNotGetManagementUrlException implements Exception {
```

File: app/lib/sharezone_plus/page/sharezone_plus_page_controller.dart [hunk 3]
```diff
@@ -105,13 +180,82 @@ class SharezonePlusPageController extends ChangeNotifier {
...
+  const CouldNotGetManagementUrl();
+}
+
+class CouldNotDetermineIsBuyingEnabled implements Exception {
```
Comment (Jonas-Sander) [inline app/lib/sharezone_plus/page/sharezone_plus_page_controller.dart:259]:
```suggestion
class CouldNotDetermineIsBuyingEnabledException implements Exception {
```

File: app/lib/sharezone_plus/subscription_service/is_buying_enabled.dart [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+// SPDX-License-Identifier: EUPL-1.2
+
+import 'dart:convert';
+
+import 'package:firebase_core/firebase_core.dart';
+import 'package:helper_functions/helper_functions.dart';
+import 'package:http/http.dart' as http;
+import 'package:retry/retry.dart';
+
+class BuyingFlagApi {
```
Comment (Jonas-Sander) [inline app/lib/sharezone_plus/subscription_service/is_buying_enabled.dart:16]:
```suggestion
class BuyingEnabledApi {
```

**Comment by Jonas-Sander** [comment]
grafik

The CORS errors come up in the preview app as it is hosted on another domain, right?

**Review by Jonas-Sander — APPROVED** [review]

**Comment by nilsreichardt** [comment]
> The CORS errors come up in the preview app as it is hosted on another domain, right?

I added CORS to the `isBuyingEnabled` enabled. This works now. But the `createStripeCheckoutSession` blocks the domain (and throws a 500 error code) because of the temporary domain of Firebase Hosting.

image

**Pull request merged**